controllers/helpers.go

```diff
@@ -777,5 +787,182 @@ func getPrismCentralClientForCluster(ctx context.Context, cluster *infrav1.Nutan
+	pcInfo, err := v3Client.V3.GetPrismCentral(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	if pcInfo.Resources == nil || pcInfo.Resources.Version == nil {
+		return "", fmt.Errorf("failed to get Prism Central version")
+	}
+
+	return *pcInfo.Resources.Version, nil
+}
+
+func detachVolumeGroupsFromVM(ctx context.Context, v4Client *prismclientv4.Client, vmUUID string) error {
+	log := ctrl.LoggerFrom(ctx)
+	vmResp, err := v4Client.VmApiInstance.GetVmById(&vmUUID)
+	if err != nil {
+		return fmt.Errorf("failed to get virtual machine: %w", err)
+	}
+
+	data := vmResp.GetData()
+	vm, ok := data.(vmconfig.Vm)
+	if !ok {
+		return fmt.Errorf("failed to cast response to VM")
+	}
+
+	volumeGroupsToDetach := make([]string, 0)
+	for _, disk := range vm.Disks {
+		backingInfo, ok := disk.GetBackingInfo().(vmconfig.ADSFVolumeGroupReference)
+		if !ok {
+			continue
+		}
+
+		volumeGroupsToDetach = append(volumeGroupsToDetach, *backingInfo.VolumeGroupExtId)
+	}
+
+	log.Info(fmt.Sprintf("detaching %d volume groups from virtual machine %s", len(volumeGroupsToDetach), vmUUID))
+
+	// Detach the volume groups from the virtual machine
+	for _, volumeGroup := range volumeGroupsToDetach {
+		log.Info(fmt.Sprintf("detaching volume group %s from virtual machine %s", volumeGroup, vmUUID))
+		body := &volumesconfig.VmAttachment{
+			ExtId: &vmUUID,
+		}
+		resp, err := v4Client.VolumeGroupsApiInstance.DetachVm(&volumeGroup, body)
+		if err != nil {
+			return fmt.Errorf("failed to detach volume group %s from virtual machine %s: %w", volumeGroup, vmUUID, err)
+		}
+
+		data := resp.GetData()
+		task, ok := data.(prismconfig.TaskReference)
+		if !ok {
+			return fmt.Errorf("failed to cast response to TaskReference")
+		}
+
+		// Wait for the task to complete
+		if _, err := waitForTaskCompletionV4(ctx, v4Client, *task.ExtId); err != nil {
+			return fmt.Errorf("failed to wait for task %s to complete: %w", *task.ExtId, err)
+		}
+	}
+
+	return nil
+}
+
+// waitForTaskCompletionV4 waits for a task to complete and returns the completion details
+// of the task. The function will poll the task status every 100ms until the task is
+// completed or the context is cancelled.
+func waitForTaskCompletionV4(ctx context.Context, v4Client *prismclientv4.Client, taskID string) ([]prismcommonconfig.KVPair, error) {
+	var data []prismcommonconfig.KVPair
+
+	if err := wait.PollUntilContextCancel(
+		ctx,
+		100*time.Millisecond,
+		true,
+		func(ctx context.Context) (done bool, err error) {
+			task, err := v4Client.TasksApiInstance.GetTaskById(utils.StringPtr(taskID))
+			if err != nil {
+				return false, fmt.Errorf("failed to get task %s: %w", taskID, err)
+			}
+
+			taskData, ok := task.GetData().(prismapi.Task)
+			if !ok {
+				return false, fmt.Errorf("unexpected task data type %[1]T: %+[1]v", task.GetData())
+			}
+
+			if taskData.Status == nil {
+				return false, nil
+			}
+
+			if *taskData.Status != prismapi.TASKSTATUS_SUCCEEDED {
```

Comment:
what if the task is still in progress. are we assuming here that it failed?

Reply:
it's a polling function. If it hasn't succeeded, it hasn't succeeded.